### ISSUE TITLE
TensorFlow Serving "version_labels" do not work as documented for the HTTP REST API 

### ISSUE THREAD

**Issue by ericmclachlan** [tensorflow/serving#1555]
### System information

- **OS Platform and Distribution**: Windows 10 and Linux Ubuntu 18.04 in WSL
- **TensorFlow Serving installed from**: Docker image (binary)
- **TensorFlow Serving version**: Docker image `tensorflow/serving:latest` (downloaded around January 20th)

### Describe the problem

At a high level, the version_labels feature appears not to work correctly. (This feature is illustrated with the "canary" vs "stable" version labels example in the TensorFlow [documentation](https://www.tensorflow.org/tfx/serving/serving_config)).

More specifically, it may only work with simple numeric version labels.

The following URL will be rejected by TensorFlow Serving: `host/v1/models/my_model/versions/1.0:predict`
with a `Malformed request...` error.

### Exact Steps to Reproduce

I believe that any kind of non-numeric version label will fail. (i.e. The "canary" vs "stable" example itself should fail.)

### Source code / logs

[http_rest_api_handler.cc](https://github.com/tensorflow/serving/blob/e085cb37b4b5c62154aa063dcbb781b24d5b3fd4/tensorflow_serving/model_servers/http_rest_api_handler.cc#L57) seems to define a regular expression used to parse the URL. Note that, after "versions", the regular expression defines an expectation for a numeric version.

```c++
prediction_api_regex_(
          R"((?i)/v1/models/([^/:]+)(?:/versions/(\d+))?:(classify|regress|predict))")
```

It's possible that I misread the code, but this interpretation agrees with an observation on [StackOverflow](https://stackoverflow.com/questions/59237402/how-to-specify-the-model-version-label-in-a-rest-api-request) that numeric versions are handled correctly whereas text version labels are not.

**Comment by rmothukuru** [comment]
@ericmclachlan,
Can you please confirm if you have tried Inference using Numeric Version Labels and using Non-Numeric Version Labels. 

And if you observed that Numeric Labels are working fine and Non-Numeric Version Labels are resulting in error, can you please provide the Commands which you have used in both the cases (Numeric and Non-Numeric), and the respective Output or the Error. Thanks!

**Comment by rmothukuru** [comment]
Also, you can look at [this article](https://medium.com/@brianschardt/how-to-setup-tensorflow-serving-for-production-3cc2abf7efa) for more information about Configuring Different Versions for Serving.

**Comment by ericmclachlan** [comment]
Hi @rmothukuru 

Thank you for the link but we already have TensorFlow Serving working in production. The problem lies specifically in trying to use the version_labels feature as described in the documentation.

I have just confirmed that sending requests like the following yields predictions:
```
http://localhost:8501/v1/models/model1/versions/1579843026:predict
```

However, using a URL like this does not:
```
http://localhost:8501/v1/models/model1/versions/test:predict
```

The only difference between these two is that this one uses the version_labels instead of the version itself.

The exact error message I receive in the POST response is:
```
{ "error": "Malformed request: POST /v1/models/model1/versions/test:predict" }
```

This is despite the model.config file containing the following version_labels definition:
```
config {
    name: 'model1'
    base_path: '/models/model1/'
    model_platform: "tensorflow"
    model_version_policy {
      specific {
        versions: 1579843026
      }
    }
    version_labels {
      key: 'test'
      value: 1579843026
    }
  }
```

I'm 99% sure this specific error message is being reported because this [line in the code](https://github.com/tensorflow/serving/blob/e085cb37b4b5c62154aa063dcbb781b24d5b3fd4/tensorflow_serving/model_servers/http_rest_api_handler.cc#L94) is failing; causing the "Malformed request" error defined in the previous few lines to be reported for this POST request.

Thanks for your help in investigating this problem.

**Comment by rmothukuru** [comment]
@ericmclachlan,
Can you please confirm that you have invoked the Tensorflow Model Server along with the Config File, as shown below. Thanks!

`tensorflow_model_server --port=8500 --rest_api_port=8501 --model_config_file=/home/configs/models.conf
`

**Comment by ericmclachlan** [comment]
I'm deploying the TensorFlow Server using docker-compose. Below is a simplied version:

```
tensorflow-servings:
    image: tensorflow/serving:latest
    ports:
      - 8500:8500
      - 8501:8501
    command:
      - --allow_version_labels_for_unavailable_models
      - --batching_parameters_file=/config/batching_parameters.txt
      - --enable_batching=true
      - --model_config_file=/config/all_models.config
      - --model_config_file_poll_wait_seconds=10
      - --monitoring_config_file=/config/monitoring_config.txt
      - --rest_api_timeout_in_ms=30000
   ```

---

This [issue](https://github.com/tensorflow/serving/issues/1413) on GitHub confirms my observation and suggests:

> Re: not being able to access the version using labels via HTTP - this is something that's not possible today (AFAIR) - only through the grpc interface can you declare labels :(

This makes me sad. The limitations of the HTTP implementation should be made more transparent. This investigation has now needlessly cost my company money; as I'm sure it will continue to do for others.

I'm not upset with you as an individual of course. But it is nonetheless a frustrating situation.

**Comment by Arnold1** [comment]
@ericmclachlan just out of curiosity - how does the  /config/batching_parameters.txt look like?

**Comment by ericmclachlan** [comment]
Hi @Arnold1 

The batching_parameters.txt looks like this:
```
max_batch_size { value: 1024 }
batch_timeout_micros { value: 100 }
num_batch_threads { value: 4 }
pad_variable_length_inputs: true
```

Please let me know if anything seems off.

**Comment by Liu-Da** [comment]
Same problem！

**Comment by ericmclachlan** [comment]
@gowthamkpr: It would probably be helpful if the [documentation](https://www.tensorflow.org/tfx/serving/serving_config) describing labels mentioned this limitation upfront.

It's less an issue of "This doesn't work" as much an issue of "I've just reimplemented REST and now I've discovered that this documented feature doesn't work unless I _don't_ use REST."

**Comment by misterpeddy** [comment]
@ericmclachlan thanks for pointing out the lack of documentation on this - I've added a note [1] clarifying that version labels don't work for REST.

@christisg how do we feel about fixing this? It's not a lot of work and I'd be happy to take it. We'd have 2 options:

1) Keep the current path [2] try to parse the version as an int, if failed, use it as a version_label.
2) Add a new path (model/<>/version_label/<>) for directing a request to a model using its version label. 

1) has the drawback of kinda odd behavior in that it implicitly disallows having version labels that can be parsed as int64 - which is currently undocumented behavior.
2) has the drawback of deviating greatly from REST principles as the path model/<>/version_label/<> URI no longer really represents a resource hosted by the server (the version_label is an attribute of a specific version of the model, not an identifier).

WDYT?

/cc @netfs in case he's thought about this before.

[1] https://github.com/tensorflow/serving/commit/9781ed1857ab0b1195107821ad0f3720872546a1
[2] https://github.com/tensorflow/serving/blob/e085cb37b4b5c62154aa063dcbb781b24d5b3fd4/tensorflow_serving/model_servers/http_rest_api_handler.cc#L57

**Comment by christisg** [comment]
Thanks @misterpeddy!

How about Option 3) keep the current path and use explicit prefix for labels, i.e models/<model_name>/versions/label=<version_label> ?
The explicit prefix makes it less error prone, and it also allows assigning numerical labels like "1" and "2" if one prefers to.

**Comment by ericmclachlan** [comment]
@misterpeddy, @christisg: I don't want to dilute the conversation too much; I just wanted to say thanks for seriously considering the suggestion. And thanks for all the work you're are doing for the community.

**Comment by misterpeddy** [comment]
Just a note that I have a preliminary implementation for the awesome idea @christisg mentioned and will resume with testing and merging it once we get some of our build breakage (due to incompatibilities with recent changes in upstream TF) under control.

**Comment by aaur0** [comment]
@misterpeddy Any update on this? Also, We have not moved to Tensorflow 2.0 yet. So, I was wondering if you guys will backport this fix to the older version of TFServing also.

**Comment by dhavalsays** [comment]
@misterpeddy Any update on this? It would be really nice to have version_labels supported in REST API.

**Comment by misterpeddy** [comment]
This was fixed in [this commit](https://github.com/tensorflow/serving/commit/3df036223b66738de1b873e9b163230fb7661cb4) - what is the remaining issue?

**Comment by dhavalsays** [comment]
@misterpeddy It doesn't seem to be working. Here are the steps for reproduction,

1) I used docker pull tensorflow/serving to pull latest docker image. My host PC is running windows 10

2) docker run -it -v C:\Code\deep-learning-keras-tf-tutorial\48_tf_serving:/48_tf_serving -p 8502:8502 --entrypoint /bin/bash tensorflow/serving   to enter into a container

3) tensorflow_model_server --rest_api_port=8502  --allow_version_labels_for_unavailable_models --model_config_file=/48_tf_serving/model.config.c

model.config.c has this,

``` 
model_config_list {

  config {
    name: 'email_model'
    base_path: '/48_tf_serving/saved_models'
    model_platform: 'tensorflow'
	model_version_policy {
	  specific {
		versions: 2
		versions: 3
	  }
	}
	version_labels {
	  key: 'stable'
	  value: 3
	}
  }
}
``` 

4) Using postman when I use this URL I get correct response back,

`http://localhost:8502/v1/models/email_model/versions/3:predict`

But when I use below it gives "Malformed request: POST /v1/models/email_model/versions/stable:predict" error

`http://localhost:8502/v1/models/email_model/versions/stable:predict`

**Comment by dhavalsays** [comment]
Hey actually I figured that my URL is wrong. It should be 'labels/stable:predict', I was using 'versions/stable:predict' I am all set. thank you.
